transports/quic/src/connection/connecting.rs

```diff
@@ -67,17 +84,131 @@ impl Connecting {
     }
 }
 
+async fn handshake(
+    connecting: quinn::Connecting,
+    mode: ConnectingMode,
+) -> Result<(PeerId, StreamMuxerBox), Error> {
+    match connecting.await {
+        Ok(connection) => {
+            let peer_id = Connecting::remote_peer_id(&connection);
+            match mode {
+                ConnectingMode::QUIC => {
+                    let muxer = Connection::new(connection);
+                    return Ok((peer_id, StreamMuxerBox::new(muxer)));
+                }
+                ConnectingMode::WebTransport(certhashes, noise_config) => {
+                    let (peer_id, muxer) = webtransport_connecting(
+                        peer_id,
+                        connection.clone(),
+                        certhashes,
+                        noise_config,
+                    )
+                    .await?;
+                    Ok((peer_id, StreamMuxerBox::new(muxer)))
+                }
+                ConnectingMode::Mixed(certhashes, noise_config) => {
+                    let res = webtransport_connecting(
+                        peer_id,
+                        connection.clone(),
+                        certhashes,
+                        noise_config,
+                    )
+                    .await;
+                    match res {
+                        Ok((peer_id, muxer)) => Ok((peer_id, StreamMuxerBox::new(muxer))),
+                        Err(WebtransportConnectingError::UnexpectedProtocol(_)) => {
+                            let muxer = Connection::new(connection);
+                            Ok((peer_id, StreamMuxerBox::new(muxer)))
+                        }
+                        Err(e) => Err(e.into()),
+                    }
+                }
+            }
+        }
+        Err(e) => {
+            return Err(Error::Connection(ConnectionError(e)));
+        }
+    }
+}
+
+async fn webtransport_connecting(
+    peer_id: PeerId,
+    connection: quinn::Connection,
+    certhashes: Vec<Multihash<64>>,
+    noise_config: libp2p_noise::Config,
+) -> Result<(PeerId, StreamMuxerBox), WebtransportConnectingError> {
+    loop {
+        let mut h3_conn = h3::server::builder()
+            .enable_webtransport(true)
+            .enable_connect(true)
+            .enable_datagram(false)
+            .max_webtransport_sessions(1)
+            .send_grease(true)
+            .build(h3_quinn::Connection::new(connection.clone()))
+            .await?;
```

Comment:
Does that mean we create a new `h3` `server` for each connection? Sorry if that's a noob question but if this is an incoming connection on a listener, couldn't we just have one server per listener?